File: 487-MaxConsecutiveOnesII/487-MaxConsecutiveOnesII.py
# Last updated: 6/2/2025, 9:24:44 PM
class Solution(object):
    def findMaxConsecutiveOnes(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # constrain 1 flip (0->1)
        # type, sliding window
        # Expand right pointer to explore.
        # Shrink left pointer only when the window violates the constraint.
        # Track window size with right - left + 1 whenever the window is valid.
        left = 0
        zero_count = 0
        max_len = 0

        for right in range(len(nums)):
            if nums[right] == 0:
                zero_count += 1

            while zero_count > 1:
                if nums[left] == 0:
                    zero_count -= 1
                left += 1

            max_len = max(max_len, right - left + 1)

        return max_len
        

File: 487-MaxConsecutiveOnesIi/487-MaxConsecutiveOnesIi.java
// Last updated: 1/22/2026, 12:12:58 PM
/*      # constrain 1 flip (0->1)
        # type, sliding window
        # Expand right pointer to explore.
        # Shrink left pointer only when the window violates the constraint.
        # Track window size with right - left + 1 whenever the window is valid.*/
class Solution {
    public int findMaxConsecutiveOnes(int[] nums) {
        int left = 0;
        int zero_count = 0; // constraint number of 0 < 2
        int max_length = 0;
        for (int right = 0;right<nums.length; right++){
            if (nums[right] == 0){
                zero_count += 1;
            };
            while (zero_count > 1){
                if (nums[left] == 0){
                    zero_count -= 1;
                };
                left +=1;

            };
            max_length = Math.max(max_length,right - left + 1);
        };
        return max_length;
    };
};